### PR TITLE
Zyx/autobackups

### LINKED ISSUE #1354: [Feature] Colony.dat purging
Currently, we store like unlimited save copies of the colonies.dat

We should limit that, either by amount or by date.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/48198756-remove-colonies-dat-after-x?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by zyxifrash**
Closes #1354 

# Changes proposed in this pull request:

Added method to prune oldest backups as per config
Added new config section with backup config parameters
Moved colony file backups to dedicated backups folder
Added timer for automated backups

Converts fixed backup command messages to translated strings

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 1]
```diff
@@ -1076,13 +1195,13 @@ private static NBTTagCompound loadNBTFromPath(@Nullable final File file)
      */
     public static void readFromNBT(@NotNull final NBTTagCompound compound, @NotNull final World world)
     {
-        if(!compound.hasKey(TAG_DISTANCE))
+        if (!compound.hasKey(TAG_DISTANCE))
         {
             Configurations.gameplay.workingRangeTownHallChunks =
-                    (int) ((Math.cos(45.0 / HALF_A_CIRCLE * Math.PI) * Configurations.gameplay.workingRangeTownHall) / BLOCKS_PER_CHUNK);
+              (int) ((Math.cos(45.0 / HALF_A_CIRCLE * Math.PI) * Configurations.gameplay.workingRangeTownHall) / BLOCKS_PER_CHUNK);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:1201]: MAJOR Assign this magic number 45.0 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 2]
```diff
@@ -992,58 +1009,160 @@ public static void onWorldLoad(@NotNull final World world)
...
-                @NotNull final File file = new File(saveDir, String.format(FILENAME_COLONY, i));
-                if (file.exists())
+                backupTimerTask.cancel();
+            }
+
+            if (numWorldsLoaded > 0 && saveNeeded)
+            {
+                saveColonies(false);
+            }
+            // todo Stop taking a gamble that the filename doesn't already exist, it's possible
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:1077]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 3]
```diff
@@ -952,6 +968,7 @@ public static void onWorldLoad(@NotNull final World world)
 
                     if (data.hasKey(TAG_NEW_COLONIES))
                     {
+                        // todo Confirm "amountOfColonies" = "max Colony ID"
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:971]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 4]
```diff
@@ -992,58 +1009,160 @@ public static void onWorldLoad(@NotNull final World world)
...
+                saveColonies(false);
+            }
+            // todo Stop taking a gamble that the filename doesn't already exist, it's possible
+            final File backupSaveLocation = getBackupSaveLocation(new Date());
+            // Just in case, try to create the parent backup directory if it doesn't exist
+            backupSaveLocation.getParentFile().mkdir();
+
+            try (FileOutputStream fos = new FileOutputStream(backupSaveLocation))
+            {
+                // todo remove hardcoded world 0
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:1084]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 5]
```diff
@@ -992,58 +1009,160 @@ public static void onWorldLoad(@NotNull final World world)
...
+     * backupDelay of 0 disables automatic backup
+     */
+    private static void scheduleColonyBackup()
     {
-        if (numWorldsLoaded > 0 && saveNeeded)
+        final int backupDelay = Configurations.administration.colonyBackupTimeMins;
+        if (backupDelay > 0)
         {
-            saveColonies(false);
+            backupTimerTask = new BackupColonyTimerTask();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:1021]: MAJOR Synchronize this lazy initialization of 'backupTimerTask' [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2444)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 62 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 55 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ColonyManager.java#L978](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L978): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [ColonyManager.java#L55](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L55): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 40 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [ColonyManager.java#L55](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L55): class "ColonyManager" has 61 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [ColonyManager.java#L345](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L345): Provide multiple methods instead of using "canDestroy" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [ColonyManager.java#L611](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L611): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [ColonyManager.java#L949](https://github.com/Minecolonies/minecolonies/blob/d36dc721dce05fee6563c3fd2281d24446ad26ee/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L949): Refactor this method to reduce its Cognitive Complexity from 31 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 6]
```diff
@@ -992,58 +1009,160 @@ public static void onWorldLoad(@NotNull final World world)
...
+    {
+        @Override
+        public void run()
         {
-            @NotNull final File saveDir = new File(DimensionManager.getWorld(0).getSaveHandler().getWorldDirectory(), FILENAME_MINECOLONIES_PATH);
-            final ZipOutputStream zos = new ZipOutputStream(fos);
+            Log.getLogger().info("Initiating scheduled backup");
+            backupColonyData();
+        }
+    }
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java:1039]: Is this concurrency save? It accesses some colony data, I feel like seeing concurrentExecution Exceptions here.

**Comment by Raycoms** [comment]
I'm a bit worried about the concurrency of the backup system, can we test this on the test server first?

**Comment by marchermans** [comment]
Personally I am not sure we need something complicated as this.
I would propose to check the files we have when we write a new backup and delete the ones we do not need.

Which prevents us from having this weird Task error.

**Comment by Raycoms** [comment]
and in the server tick we can check each %backupinc==0 if we need to make
one

2018-03-26 16:44 GMT+01:00 Marc Hermans <notifications@github.com>:

**Comment by marchermans** [comment]
Yep.

**Comment by Raycoms** [comment]
Since this was never adapted to cope with multiple threads and I'll rewrite this soon to use capabilities as well. I can close this for now.

**Pull request closed without merging**